### PR TITLE
[TIR][USMP] Integrating USMP to AoT Executor

### PR DESCRIPTION
**Pull request by manupa-arm**
RFC : https://github.com/apache/tvm-rfcs/blob/main/rfcs/0009_Unified_Static_Memory_Planning.md

This commit integrates the USMP to AoT executor codegen as a Pass.
It also introduces two PassContext options to disable USMP as well as
StorageRewrite passes. Currently it will only support U1 usecase where
TVM internally generates a workspace buffer. The creation of constant
buffers will be added when we have tir.allocate_const support.

### CODE REVIEW

**Comment by manupa-arm** [comment]
Putting this to draft as it containts the contents of the dependent PRs (See above)

File: include/tvm/tir/usmp/transform.h [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+namespace tir {
+namespace usmp {
+namespace transform {
+
+using Pass = tvm::transform::Pass;
+
+/*!
+ * \brief Convert the analyzed PoolAllocation to offsets from pool variables
+ *
+ * This pass would convert the IRModule that contains all PrimFuncs that contains
```
Comment (mbs-octoml) [inline include/tvm/tir/usmp/transform.h:40]: nit: run on sentence is hard to parse, maybe split into both adding base address argument and converting allocs to base address + offset.

File: include/tvm/tir/usmp/transform.h [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+ *
+ * \return the pass
+ */
+TVM_DLL Pass ConvertPoolAllocationsToOffsets(const Map<tir::Stmt, PoolAllocation>& pool_allocations,
+                                             Bool emit_tvmscript_printable = Bool(false));
+
+/*!
+ * \brief Assign PoolInfo objects to tir.allocate nodes depending on the PrimFunc's target
+ *
+ * This pass would assign PoolInfo objects to tir.allocate nodes depending on the each target
```
Comment (mbs-octoml) [inline include/tvm/tir/usmp/transform.h:52]: nit: maybe assign default PollInfo objects to allocate nodes not otherwise annotated,  depending on pool info
supplied for each target.

File: include/tvm/tir/usmp/utils.h [hunk 3]
```diff
@@ -59,22 +60,26 @@ struct PoolInfoNode : public Object {
   Integer size_hint_bytes;
   /*! \brief The accessibility from each Target*/
   Map<Target, String> target_access;  // 'rw' or 'ro'
+  /*! \brief Whether pool is internally generated*/
+  Bool is_internal = Bool(false);
```
Comment (mbs-octoml) [inline include/tvm/tir/usmp/utils.h:64]: out of curiosity any reason to use the TIR Bool node over just bool? FFI is all perfectly happy with plain old bool.

Comment (mbs-octoml) [reply include/tvm/tir/usmp/utils.h:64]: nit: one extra sentence 'Internally generated pools are needed for ... and treated differently by ...' would help.

File: include/tvm/tir/usmp/utils.h [hunk 4]
```diff
@@ -294,6 +319,12 @@ namespace attr {
  */
 static constexpr const char* kPoolArgs = "pool_args";
 
+/*!
+ * \brief This is a BaseFunc attribute to indicate which input var represent
```
Comment (mbs-octoml) [inline include/tvm/tir/usmp/utils.h:323]: copy pasta from above, what is this bound to?

File: include/tvm/tir/transform.h [hunk 5]
```diff
@@ -484,6 +484,13 @@ TVM_DLL Pass MergeDynamicSharedMemoryAllocations();
  */
 TVM_DLL Pass ConvertForLoopsToSerial();
 
+/*!
+ * \brief This is the unified static memory planner pass that will
+ * plan for memory intra- and inter- PrimFuncs together.
+ * \return The pass.
```
Comment (mbs-octoml) [inline include/tvm/tir/transform.h:490]: nit: state reqs all globals to be PrimFuncs, inc main.

File: src/tir/usmp/unified_static_memory_planner.cc [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+               {"greedy_by_conflicts", algo::GreedyByConflicts}};
+
+IRModule PlanMemory(const IRModule& mod, String algo) {
+  VLOG(1) << "workspace required = " << CalculateModuleWorkspaceSize(mod);
+  PrimFunc main_func = Downcast<PrimFunc>(mod->Lookup(::tvm::runtime::symbol::tvm_run_func_suffix));
+  BufferInfoAnalysis buffer_info_analysis = ExtractBufferInfo(main_func, mod);
+  Array<BufferInfo> buffer_info_arr =
+      CreateArrayBufferInfo(buffer_info_analysis->buffer_info_stmts);
+  Map<BufferInfo, PoolAllocation> buffer_info_pool_allocations =
+      algorithms[algo](buffer_info_arr, buffer_info_analysis->memory_pressure);
```
Comment (mbs-octoml) [inline src/tir/usmp/unified_static_memory_planner.cc:58]: suggest a diagnostic for invalid algorithm tag

**Review by mbs-octoml — CHANGES_REQUESTED** [review]
Will need to spend more time on aot_executor_codegen.cc and source_module.cc, or perhaps @areusch can help.

File: include/tvm/tir/usmp/transform.h (hunk 1 above)
Comment (manupa-arm) [reply include/tvm/tir/usmp/transform.h]: Done

File: include/tvm/tir/usmp/transform.h (hunk 2 above)
Comment (manupa-arm) [reply include/tvm/tir/usmp/transform.h]: Done

File: include/tvm/tir/usmp/utils.h (hunk 3 above)
Comment (manupa-arm) [reply include/tvm/tir/usmp/utils.h]: Done

File: include/tvm/tir/transform.h (hunk 5 above)
Comment (manupa-arm) [reply include/tvm/tir/transform.h:491]: Done

File: include/tvm/tir/usmp/utils.h (hunk 4 above)
Comment (manupa-arm) [reply include/tvm/tir/usmp/utils.h]: Yes. good catch! fixed

File: src/tir/usmp/unified_static_memory_planner.cc (hunk 6 above)
Comment (manupa-arm) [reply src/tir/usmp/unified_static_memory_planner.cc:60]: Added that!

File: src/target/source/codegen_source_base.cc [hunk 7]
```diff
@@ -113,5 +113,15 @@ void CodeGenSourceBase::EndScope(int scope_id) {
   indent_ -= 2;
 }
 
+std::string CodeGenSourceBase::SanitiseName(std::string name) {
```
Comment (Mousius) [inline src/target/source/codegen_source_base.cc:116]: Can we use https://github.com/apache/tvm/blob/main/src/relay/backend/name_transforms.cc#L87-L95 ?

File: include/tvm/tir/usmp/algo/algo.h [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/*!
+ * \file tir/usmp/algo/algo.h
```
Comment (Mousius) [inline include/tvm/tir/usmp/algo/algo.h:21]: nit: is there a reason to not just have an `algorithms.h` instead of an `algo.h` nested in an  `algo` folder. `algo` is also unnecessarily ambiguous.

File: src/relay/backend/aot_executor_codegen.cc [hunk 9]
```diff
@@ -645,32 +632,120 @@ class AOTExecutorCodegen : public MixedModeVisitor {
...
+   */
+  IRModule PlanMemoryLoweredModule(const IRModule& mod) {
+    transform::PassContext pass_ctx = transform::PassContext::Current();
+    bool enable_usmp = pass_ctx->GetConfig<Bool>("tir.usmp.enable", Bool(false)).value();
+
+    IRModule lowered_mod = mod->ShallowCopy();
+    Executor executor_config = mod->GetAttr<Executor>(tvm::attr::kExecutor).value();
+    Integer workspace_byte_alignment =
+        executor_config->GetAttr<Integer>("workspace-byte-alignment").value_or(16);
+    if (enable_usmp) {
```
Comment (Mousius) [inline src/relay/backend/aot_executor_codegen.cc:701]: Would be easier to read these two blocks in their own functions, such as:
```c++
    if (enable_usmp) {
       return PlanMemoryLoweredModuleWithUSMP(mod);
    } else {
       return PlanMemoryLoweredModuleWithoutUSMP(mod);
    }
```

File: src/tir/transforms/make_unpacked_api.cc [hunk 10]
```diff
@@ -57,33 +57,26 @@ PrimFunc MakeUnpackedAPI(PrimFunc&& func) {
   const Stmt nop = Evaluate(0);
   std::vector<Stmt> device_init;
 
-  // Create arg to buffer binder
-  std::unordered_map<const VarNode*, PrimExpr> vmap;
-  ArgBinder binder(&vmap);
```
Comment (Mousius) [inline src/tir/transforms/make_unpacked_api.cc:62]: This is a great improvement :smile_cat:

File: tests/python/relay/aot/test_c_device_api.py [hunk 11]
```diff
@@ -136,44 +136,29 @@ def test_device_api_hooks_unpacked_api(device_api_main_func):
...
-        str(main_func.body[0][0].value)
-        == "@tir.call_extern("
-        + '"TVMDeviceEthosUActivate",'
-        + " device_context_ethos_u: handle,"
-        + " dtype=int32)"
+        str(main_func.body[0])
+        == "tir.call_extern(" + '"TVMDeviceEthosUActivate",' + " device_context_ethos_u)\n"
     )
     # Open Device
+    print(str(main_func.body[1]))
```
Comment (Mousius) [inline tests/python/relay/aot/test_c_device_api.py:143]: Is this meant to still be here? If so it would make sense to print the entire `body` before any asserts?

File: src/target/source/source_module.cc [hunk 12]
```diff
@@ -26,6 +26,9 @@
 #include <tvm/runtime/ndarray.h>
 #include <tvm/runtime/packed_func.h>
 #include <tvm/runtime/registry.h>
+#include <tvm/tir/function.h>
+#include <tvm/tir/stmt_functor.h>
+#include <tvm/tir/transform.h>
```
Comment (Mousius) [inline src/target/source/source_module.cc:31]: It's likely that these are the reason why your Windows CI builds are failing as iirc `source_module.cc` only gets linked with the runtime, not the entirety of TVM?

**Review by Mousius — CHANGES_REQUESTED** [review]
Just had a quick look at this, mostly looks super exciting with a few bits to highlight but also hopefully help fix the Windows CI :smile_cat:

File: src/tir/usmp/utils.cc [hunk 13]
```diff
@@ -168,6 +174,20 @@ Array<BufferInfo> CreateArrayBufferInfo(const Map<BufferInfo, Stmt>& buffer_info
   return ret;
 }
 
+void PrintConflicts(const Array<BufferInfo>& bi_arr) {
+  for (const auto& bi : bi_arr) {
```
Comment (areusch) [inline src/tir/usmp/utils.cc:178]: want to either log each conflict separately or wrap this function in a log level check? (or, is this function called?)

File: tests/python/relay/aot/test_c_device_api.py [hunk 14]
```diff
@@ -231,26 +216,29 @@ def test_without_device_api_unpacked_api(non_device_api_main_func):
     """Test a graph without the Device API with the unpacked internal calls"""
 
     main_func = non_device_api_main_func(interface_api="c", use_unpacked_api=True)
-
+    print(str(main_func.body))
```
Comment (areusch) [inline tests/python/relay/aot/test_c_device_api.py:219]: nit: delete

**Review by areusch — COMMENTED** [review]
thanks @manupa-arm this looks pretty good! very exciting to see this work progress to the point of integration. left a few comments here and there for my understanding, and a couple small cleanups

File: src/target/source/codegen_source_base.cc (hunk 7 above)
Comment (manupa-arm) [reply src/target/source/codegen_source_base.cc]: Ah, I missed this one. Yes its better to use that. corrected

File: src/tir/transforms/make_unpacked_api.cc (hunk 10 above)
Comment (manupa-arm) [reply src/tir/transforms/make_unpacked_api.cc:62]: Thanks :) !

File: tests/python/relay/aot/test_c_device_api.py (hunk 11 above)
Comment (manupa-arm) [reply tests/python/relay/aot/test_c_device_api.py]: You got me!. Removed.

File: src/target/source/source_module.cc (hunk 12 above)
Comment (manupa-arm) [reply src/target/source/source_module.cc]: Thanks! lets see. I was scratching my head over this.

File: src/relay/backend/aot_executor_codegen.cc [hunk 15]
```diff
@@ -859,20 +922,28 @@ class AOTExecutorCodegen : public MixedModeVisitor {
...
-    } else {
-      VLOG(1) << "adding main into new module for host target";
-      ret.lowered_funcs.Set(target_host_, mod_run);
+    Map<tir::Var, tir::usmp::AllocatedPoolInfo> pool_var_info;
+    std::vector<tir::Var> pool_vars;
+    tir::PrimFunc tir_main_func =
+        Downcast<tir::PrimFunc>(lowered_mod->Lookup(::tvm::runtime::symbol::tvm_run_func_suffix));
+    Optional<Array<tir::usmp::AllocatedPoolInfo>> allocated_pool_infos =
+        tir_main_func->GetAttr<Array<tir::usmp::AllocatedPoolInfo>>(tvm::attr::kPoolArgs);
+    int main_workspace_size = 0;
```
[the remark this answers is not in the document]
Comment (manupa-arm) [reply src/relay/backend/aot_executor_codegen.cc]: Removed. not needed really :)

File: src/relay/backend/aot_executor_codegen.cc [hunk 16]
```diff
@@ -645,32 +632,120 @@ class AOTExecutorCodegen : public MixedModeVisitor {
...
+    }
+  }
+
+  /*!
+   * brief This function is a wrapper to run memory planning
+   * followed by recording the latest workspaces required.
+   */
+  IRModule PlanMemoryLoweredModule(const IRModule& mod) {
+    transform::PassContext pass_ctx = transform::PassContext::Current();
+    bool enable_usmp = pass_ctx->GetConfig<Bool>("tir.usmp.enable", Bool(false)).value();
```
[the remark this answers is not in the document]
Comment (manupa-arm) [reply src/relay/backend/aot_executor_codegen.cc]: made into constants.

File: src/relay/backend/aot_executor_codegen.cc [hunk 17]
```diff
@@ -645,32 +632,120 @@ class AOTExecutorCodegen : public MixedModeVisitor {
...
+    IRModule lowered_mod = mod->ShallowCopy();
+    Executor executor_config = mod->GetAttr<Executor>(tvm::attr::kExecutor).value();
+    Integer workspace_byte_alignment =
+        executor_config->GetAttr<Integer>("workspace-byte-alignment").value_or(16);
+    if (enable_usmp) {
+      lowered_mod = tir::transform::UnifiedStaticMemoryPlanner()(lowered_mod);
+      // Update workspace size based on the pool allocations.
+      Optional<Array<tir::usmp::AllocatedPoolInfo>> allocated_pool_infos =
+          lowered_mod->GetAttr<Array<tir::usmp::AllocatedPoolInfo>>(tvm::attr::kPoolArgs);
+      int main_workspace_size = 0;
```
[the remark this answers is not in the document]
Comment (manupa-arm) [reply src/relay/backend/aot_executor_codegen.cc]: Changed to report target based workspace size through analyzing pool's target accessibility.

File: src/relay/backend/aot_executor_codegen.cc [hunk 18]
```diff
@@ -645,32 +632,120 @@ class AOTExecutorCodegen : public MixedModeVisitor {
...
+        for (const tir::usmp::AllocatedPoolInfo& allocated_pool_info :
+             allocated_pool_infos.value()) {
+          main_workspace_size += allocated_pool_info->allocated_size->value;
+        }
+      }
+      for (const auto& kv : function_metadata_) {
+        if (lowered_mod->ContainGlobalVar(kv.first) &&
+            lowered_mod->Lookup(kv.first)->IsInstance<tir::PrimFuncNode>()) {
+          tir::PrimFunc pfunc = Downcast<tir::PrimFunc>(lowered_mod->Lookup(kv.first));
+          Target tgt = pfunc->GetAttr<Target>(tvm::attr::kTarget).value();
```
[the remark this answers is not in the document]
Comment (manupa-arm) [reply src/relay/backend/aot_executor_codegen.cc]: Could we do it in a follow on ? we need MLF reporting structure to be adjusted too before we do that too.

File: src/target/source/source_module.cc [hunk 19]
```diff
@@ -197,45 +202,161 @@ class CSourceCrtMetadataModuleNode : public runtime::ModuleNode {
...
   }
 
+  String GenerateDLTensorStructWrapper(String reference_arg) {
+    code_ << "DLTensor " << reference_arg << "_dlt = {\n";
+    code_ << ".data = &" << reference_arg << "\n";
+    code_ << "};\n";
+    code_ << "TVMValue " << reference_arg << "_tvmv = {\n";
+    code_ << ".v_handle = &" << reference_arg << "_dlt\n";
+    code_ << "};\n";
+    return reference_arg + "_tvmv";
```
[the remark this answers is not in the document]
Comment (manupa-arm) [reply src/target/source/source_module.cc]: Done.

File: src/target/source/source_module.cc [hunk 20]
```diff
@@ -197,45 +202,161 @@ class CSourceCrtMetadataModuleNode : public runtime::ModuleNode {
           << "}\n";
   }
 
+  String GenerateDLTensorStructWrapper(String reference_arg) {
+    code_ << "DLTensor " << reference_arg << "_dlt = {\n";
```
[the remark this answers is not in the document]
Comment (manupa-arm) [reply src/target/source/source_module.cc]: Done.

File: src/target/source/source_module.cc [hunk 21]
```diff
@@ -197,45 +202,161 @@ class CSourceCrtMetadataModuleNode : public runtime::ModuleNode {
...
     code_ << "(void* args, void* type_code, int num_args, void* out_value, void* "
              "out_type_code, void* resource_handle) {\n";
     code_ << "return " << run_func << "(";
-    for (unsigned int i = 0; i < metadata_->inputs.size(); ++i) {
-      code_ << "((DLTensor*)(((TVMValue*)args)[" << i << "].v_handle))[0].data,";
+
+    {
+      std::stringstream call_args_ss;
+      for (unsigned int i = 0; i < metadata_->inputs.size(); ++i) {
+        call_args_ss << "((DLTensor*)(((TVMValue*)args)[" << i << "].v_handle))[0].data,";
```
[the remark this answers is not in the document]
Comment (manupa-arm) [reply src/target/source/source_module.cc:279]: Being consistent with codegen_c, unless I missed a change to that.

File: src/tir/usmp/utils.cc (hunk 13 above)
Comment (manupa-arm) [reply src/tir/usmp/utils.cc]: Removed, it is not called.

File: tests/python/relay/aot/test_c_device_api.py (hunk 14 above)
Comment (manupa-arm) [reply tests/python/relay/aot/test_c_device_api.py]: Removed.

**Comment by manupa-arm** [comment]
A friendly ping @Mousius @areusch ...

File: src/target/source/codegen_source_base.h [hunk 22]
```diff
@@ -52,6 +52,18 @@ class CodeGenSourceBase {
    * \param value The constant value.
    */
   void MarkConst(std::string value);
+  /*!
+   * Print Type represetnation of type t.
```
Comment (Mousius) [inline src/target/source/codegen_source_base.h:56]:
```suggestion
   * Print Type representation of type t.
```

File: src/target/source/codegen_source_base.h [hunk 23]
```diff
@@ -52,6 +52,18 @@ class CodeGenSourceBase {
...
+   * \param t The type representation.
+   * \param os The stream to print the ctype into
+   */
+  virtual void PrintType(DataType t, std::ostream& os);  // NOLINT(*)
+  /*!
+   * Print Type represetnation of type type.
+   * \param type The type representation.
```
Comment (Mousius) [inline src/target/source/codegen_source_base.h:63]: Can we use the same names in both of these blocks? i.e. `type type` or `type t` ? 

Also minor typo:
```suggestion
   * Print Type representation of type type.
   * \param type The type representation.
```

**Review by Mousius — CHANGES_REQUESTED** [review]
Just minor comments from me, I think the rest of my comments were dealt with :smile_cat:

File: src/target/source/codegen_source_base.h (hunk 23 above)
Comment (manupa-arm) [reply src/target/source/codegen_source_base.h]: Done

File: src/target/source/codegen_source_base.h (hunk 22 above)
Comment (manupa-arm) [reply src/target/source/codegen_source_base.h]: Done!

File: src/relay/backend/aot_executor_codegen.cc (hunk 18 above)
[the remark this answers is not in the document]
Comment (areusch) [reply src/relay/backend/aot_executor_codegen.cc]: sure

File: src/target/source/codegen_source_base.cc [hunk 24]
```diff
@@ -113,5 +113,60 @@ void CodeGenSourceBase::EndScope(int scope_id) {
...
+  if (type.is_handle()) {
+    os << "void*";
+    return;
+  }
+  if (type.is_float()) {
+    if (type.bits() == 32) {
+      os << "float";
+      return;
+    }
+    if (type.bits() == 64) {
```
Comment (areusch) [inline src/target/source/codegen_source_base.cc:127]: should this have an `else CHECK(false)`? same question elsewhere

**Review by areusch — APPROVED** [review]
thanks @manupa-arm!

File: src/relay/backend/aot_executor_codegen.cc (hunk 18 above)
[the remark this answers is not in the document]
Comment (mbs-octoml) [reply src/relay/backend/aot_executor_codegen.cc]: Yeah, no need to get hung up on that -- the target attribute is the official carrier.

File: src/target/source/codegen_source_base.cc (hunk 24 above)
Comment (manupa-arm) [reply src/target/source/codegen_source_base.cc:127]: I was just moving the already existing implementation to the base class :).
Moreover, there was a LOG(FATAL) at the bottom that should cover this case anyway.

**Comment by manupa-arm** [comment]
@Mousius do you think we could merge this one ?

**Review by Mousius — APPROVED** [review]
> @Mousius do you think we could merge this one ?

Definitely @manupa-arm! Thanks for the reviews from @mbs-octoml and @areusch!
